lib/transports/sshTransport.js

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) International Business Machines Corp. 2019
+// All Rights Reserved
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
+// associated documentation files (the "Software"), to deal in the Software without restriction,
+// including without limitation the rights to use, copy, modify, merge, publish, distribute,
+// sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+
+// The above copyright notice and this permission notice shall be included in all copies or
+// substantial portions of the Software.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
+// NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+function sshCall(config, xmlIn, done) {
+  const { Client } = require('ssh2');
+
+  const {
+    verbose = false,
+  } = config;
+
+  let xmlOut = '';
+  const xmlBuffer = Buffer.from(xmlIn);
+
+  const client = new Client();
+
+  // subscribe for events
+  client.on('error', (error) => {
+    client.end();
+    done(error, null);
+  });
+
+  client.on('close', () => {
+    if (verbose) {
+      console.log('SSH Client has closed');
+    }
+  });
+
+  client.on('end', () => {
+    if (verbose) {
+      console.log('SSH Client has ended');
+    }
+  });
+
+  client.on('ready', () => {
+    if (verbose) {
+      console.log('SSH Client is ready');
+    }
+    client.exec('/QOpenSys/pkgs/bin/xmlservice-cli', (error, stream) => {
+      if (error) {
+        done(error, null);
+        client.end();
+        return;
+      }
+      stream.stdout.on('close', (code) => {
+        if (verbose) {
+          console.log(`stdout has closed with code: ${code}`);
+        }
+        done(null, xmlOut);
+        client.end();
+      });
+
+      stream.stdin.on('end', () => {
+        if (verbose) {
+          console.log('stdin has ended');
+        }
+      });
+
+      stream.stdout.on('data', (data) => {
+        xmlOut += data.toString();
+        if (verbose) {
+          console.log(`STDOUT:\n${data}`);
+        }
+      });
+
+      stream.stderr.on('data', (data) => {
+        if (verbose) {
+          console.log(`STDERR:\n${data}`);
+        }
+      });
+
+      stream.stdin.write(xmlBuffer);
+      stream.stdin.end();
+    });
+  });
```

Comment:
Seems to me we are not handling the `exit` event, which would have the return code. If `xmlservice-cli` returns a non-zero return to indicate failure, we would return a false positive since we never received that here.

https://www.npmjs.com/package/ssh2#channel

```
For exec():

    An exit event may (the SSH2 spec says it is optional) be emitted when the process finishes. If the process finished normally, the process's return value is passed to the exit callback. If the process was interrupted by a signal, the following are passed to the exit callback: null, < string >signalName, < boolean >didCoreDump, < string >description.

    If there was an exit event, the close event will be passed the same arguments for convenience.
```

Reply:
@kadler Good call I'm going to add:

```javascript
     stream.on('exit', (code, signal, didCoreDump, description) => {
        if (verbose) {
          console.log(`Stream exit code: ${code}`);

          if (signal) {
            console.log(`Signal: ${signal}, Description: ${description}`);
          }
        }

        if (code !== 0) {
          client.end();
          done(`xmlserivce-cli exited abnormally`, null);
          return;
        }

        done(null, xmlOut);
        client.end();
      });

```